### DISCUSSION TITLE
Compared to ts-rest

### DISCUSSION THREAD

**Discussion by gunzip** [ecyrbe/zodios#342] [Announcements]
Hi 👋 I appreciate all your hard work. I came across this issue at https://github.com/ts-rest/ts-rest/issues/14 and am curious: what are the main differences between the two tools from the perspective of the end user? Do you advise using Zodios for some scenarios and ts-rest for others?

**Comment by ecyrbe** [comment]
for now i can show some of the different features we have :
| feature                              | side          | zodios      | ts-rest | trpc                      |
| ------------------------------------ | ------------- | ----------- | ------- | ------------------------- |
| contract based                       | client/server | yes ✅       | yes ✅   | no ❌                      |
| work outside mono repo               | client/server | yes ✅       | yes ✅   | no ❌                      |
| typesafe                             | client/server | yes ✅       | yes ✅   | yes ✅                     |
| validation with any library (not just zod)          | client/server | yes ✅ (v11) | no ❌    | partial ✅                 |
| validation on client                 | client        | yes✅        | yes ✅   | no ❌                      |
| plugins                              | client        | yes ✅       | no ❌    | partial ✅ (links)         |
| typesafe error handling              | client        | yes ✅       | yes ✅   | yes ✅                     |
| fetch native support                 | client        | yes ✅ (v11) | yes ✅   | yes ✅                     |
| axios native support                 | client        | yes ✅       | no ❌    | no ❌                      |
| react client                         | client        | yes ✅       | yes ✅   | yes ✅                     |
| solid client                         | client        | yes ✅       | yes ✅   | yes ✅                     |
| testing/mocking |client        | yes ✅  (v11)     | no ❌    | no ❌                      |
| automatic react-query key management | client        | yes ✅       | no ❌    | yes ✅                     |
| automatic batching | client | no ❌        |no ❌                      | yes ✅   | 
| nextjs support                       | client        | yes ✅       | yes ✅   | yes ✅                     |
| validation on server                 | server        | yes ✅       | yes ✅   | yes ✅                     |
| typesafe error handling              | server        | yes ✅       | yes ✅   | yes ✅                     |
| nextjs support                       | server        | yes ✅       | yes ✅   | yes ✅                     |
| nestjs support                       | server        | no ❌        | yes ✅   | no ❌                      |
| server written like in express       | server        | yes ✅       | no ❌    | no ❌                      |
| typesafe middleware support          | server        | yes ✅       | no ❌    | yes ✅                     |
| typesafe context support             | server        | yes ✅       | no ❌    | yes ✅                     |
| compatible with serverless           | server        | yes ✅       | yes ✅   | yes ✅                     |
| openapi support                      | server        | yes ✅       | yes ✅   | partial ✅  (external lib) |

**Reply by vedxyz** [reply]
Honestly, I don't think it would be a bad idea to add a section with this kind of comparison between the three libraries to the README. I found it a bit hard to differentiate the three definitively without this table.

That said, I think Zodios is indeed what I'm in need of right now :).

**Comment by ivan-kleshnin** [comment]
Nice table, thanks 👍 My 5 cents: bundled Axios is a downside for anyone who doesn't use this library (like myself). More important criteria to compare would be a bundler of choice or ESM support. From my experience with `trpc`, `next-fetch` and `ts-rest` integration is the major pain point. When NextJS wants Turbopack and the library needs Vite or Babel... you're out of luck 😱
